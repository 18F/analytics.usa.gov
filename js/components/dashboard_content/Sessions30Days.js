import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";
import buildTimeSeries from "../../lib/chart_helpers/timeseries";
import formatters from "../../lib/chart_helpers/formatters";

/**
 * Retrieves the 30 days sessions report from the passed data URL and creates a
 * visualization for the count of sessions visiting sites for the current agency
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @returns {import('react').ReactElement} The rendered element
 */
function Sessions30Days({ dataHrefBase }) {
  const jsonDataURL = `${dataHrefBase}/daily-sessions-30-days.json`;
  const csvDataURL = `${dataHrefBase}/daily-sessions-30-days.csv`;
  const ref = useRef(null);
  const [sessionData, setSessionData] = useState(null);

  useEffect(() => {
    const initSessionsChart = async () => {
      if (!sessionData) {
        const data = await DataLoader.loadDailyReportJSON(jsonDataURL);
        await setSessionData(data);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder
          .setElement(ref.current)
          .setData(sessionData)
          .setTransformer((data) => data)
          .setRenderer((svg, data) => {
            const days = data.data;
            days.forEach((d) => {
              d.visits = +d.visits;
            });

            const y = function (d) {
              return d.visits;
            };

            const series = buildTimeSeries()
              .series([data.data])
              .y(y)
              .label((d) => formatters.formatDate(d.date))
              .title(
                (d) =>
                  `${formatters.addCommas(d.visits)} visits during the day of ${d.date}`,
              );

            series.xScale().domain(d3.range(0, days.length + 1));

            series.yScale().domain([0, d3.max(days, y)]);

            series.yAxis().tickFormat(formatters.formatVisits());

            svg.call(series);
          })
          .build();
      }
    };
    initSessionsChart().catch(console.error);
  }, [sessionData]);

  return (
    <>
      <div className="chart__title">
        Daily Sessions
        <a href={csvDataURL} aria-label="daily-sessions-30-days.csv">
          <svg
            className="usa-icon margin-bottom-neg-05 margin-left-05"
            aria-hidden="true"
            focusable="false"
            role="img"
          >
            <use xlinkHref="/assets/uswds/img/sprite.svg#file_present"></use>
          </svg>
        </a>
      </div>
      <div className="chart__time-series__container" ref={ref}>
        <svg className="data chart__time-series"></svg>
      </div>
    </>
  );
}

Sessions30Days.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default Sessions30Days;
