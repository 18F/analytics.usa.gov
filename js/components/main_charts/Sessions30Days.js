import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";
import buildTimeSeries from "../../lib/chart_helpers/timeseries";
import formatters from "../../lib/chart_helpers/formatters";

/**
 * Retrieves the 30 days sessions report from the passed data URL and creates a
 * visualization for the count of sessions visiting sites for the current agency
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @returns {import('react').ReactElement} The rendered element
 */
function Sessions30Days({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/sessions-over-30-days.json`;
  const ref = useRef(null);
  const [sessionData, setSessionData] = useState(null);

  useEffect(() => {
    const initSessionsChart = async () => {
      if (!sessionData) {
        const data = await DataLoader.loadJSON(dataURL);
        await setSessionData(data);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder
          .setElement(ref.current)
          .setData(sessionData)
          .setTransformer((data) => data)
          .setRenderer((svg, data) => {
            const days = data.data;
            days.forEach((d) => {
              d.visits = +d.visits;
            });

            const y = function (d) {
              return d.visits;
            };

            const series = buildTimeSeries()
              .series([data.data])
              .y(y)
              .label((d) => formatters.formatDate(d.date))
              .title(
                (d) =>
                  `${formatters.addCommas(d.visits)} visits during the day of ${d.date}`,
              );

            series.xScale().domain(d3.range(0, days.length + 1));

            series.yScale().domain([0, d3.max(days, y)]);

            series.yAxis().tickFormat(formatters.formatVisits());

            svg.call(series);
          })
          .build();
      }
    };
    initSessionsChart().catch(console.error);
  }, [sessionData]);

  return (
    <div
      id="time_series"
      className="section__chart"
      data-refresh="15"
      ref={ref}
    >
      <svg className="data time-series"></svg>
    </div>
  );
}

Sessions30Days.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default Sessions30Days;
