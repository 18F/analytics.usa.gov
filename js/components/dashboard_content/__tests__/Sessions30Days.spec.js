import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import { delay } from "../../../../spec/support/test_utilities";
import DataLoader from "../../../lib/data_loader";
import Sessions30Days from "../Sessions30Days";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadDailyReportJSON: jest.fn(),
}));

describe("Sessions30Days", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadDailyReportJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <Sessions30Days dataHrefBase="http://www.example.com/data/" />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    beforeEach(async () => {
      data = {
        name: "sessions-over-30-days",
        query: {},
        meta: {
          name: "Sessions Over 30 Days",
          description: "Visits for all sites over 30 days.",
        },
        data: [
          {
            date: "2023-12-17",
            visits: "29515461",
          },
          {
            date: "2023-12-18",
            visits: "51017053",
          },
          {
            date: "2023-12-19",
            visits: "48308250",
          },
        ],
        totals: {
          visits: 1186512188,
        },
        taken_at: "2024-01-16T19:50:29.099Z",
      };

      DataLoader.loadDailyReportJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <Sessions30Days dataHrefBase="http://www.example.com/data/" />,
      );
      await waitFor(() => screen.getByText("12/17"));
      // Wait for barchart transition animation to complete (200 ms, set in
      // js/lib/chart_helpers/barchart.js)
      await delay(300);
    });

    // The timeseries data intermittently has slightly different values for
    // y values, height, opacity of ticks. So rather than using a snapshot of
    // the fully rendered component, instead verify that the data is loaded to
    // the screen by searching for text.
    it("renders a component with data loaded", async () => {
      expect(
        await screen.findByText(
          "29,515,461 visits during the day of 2023-12-17",
        ),
      ).not.toBeNull();
      expect(
        await screen.findByText(
          "51,017,053 visits during the day of 2023-12-18",
        ),
      ).not.toBeNull();
      expect(
        await screen.findByText(
          "48,308,250 visits during the day of 2023-12-19",
        ),
      ).not.toBeNull();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadDailyReportJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <Sessions30Days dataHrefBase="http://www.example.com/data/" />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("logs the error to console", () => {
      expect(console.error).toHaveBeenCalledWith(error);
    });
  });
});
