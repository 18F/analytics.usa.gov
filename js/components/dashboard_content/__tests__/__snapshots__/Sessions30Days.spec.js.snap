// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sessions30Days when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <div
    class="chart__time-series__container"
  >
    <svg
      class="data chart__time-series"
    />
  </div>
</DocumentFragment>
`;

exports[`Sessions30Days when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <div
    class="chart__time-series__container"
  >
    <svg
      class="data chart__time-series"
    />
  </div>
</DocumentFragment>
`;
